python/lsst/ap/association/apDataProductMaker.py

```diff
@@ -0,0 +1,196 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+"""Classes for taking science pipeline outputs and creating data products for
+use in ap_association and the prompt-products database (PPDB).
+"""
+
+__all__ = ["APDataMapperConfig", "APDataMapperTask",
+           "DiaSourceMapperConfig", "DiaSourceMapperTask"]
+
+import lsst.afw.table as afwTable
+from lsst.daf.base import DateTime
+import lsst.pipe.base as pipeBase
+import lsst.pex.config as pexConfig
+import lsst.afw.image as afwImage
+from .afwUtils import make_minimal_dia_source_schema
+
+
+class APDataMapperConfig(pexConfig.Config):
+    """Configuration for the generic APDMMapper class.
+    """
+    copyColumns = pexConfig.DictField(
+        keytype=str,
+        itemtype=str,
+        doc="Mapping of input SciencePipelines columns to output DPDD "
+            "columns.",
+        default={"id": "id",
+                 "parent": "parent",
+                 "coord_ra": "coord_ra",
+                 "coord_dec": "coord_dec"}
+    )
+
+
+class APDataMapperTask(pipeBase.Task):
+    """Generic mapper class for copying values from a science pipelines catalog
+    into a product for use in ap_association or the PPDB.
+    """
+    ConfigClass = APDataMapperConfig
+    _DefaultName = "apDataMapperTask"
+
+    def __init__(self, inputSchema, outputSchema, **kwargs):
+        pipeBase.Task.__init__(self, **kwargs)
+        self.inputSchema = inputSchema
+        self.outputSchema = outputSchema
+
+        self.mapper = afwTable.SchemaMapper(inputSchema, outputSchema)
+
+        for inputName, outputName in self.config.copyColumns.items():
+            self.mapper.addMapping(
+                self.inputSchema.find(inputName).key,
+                outputName,
+                True)
+
+    def run(self, inputCatalog, exposure=None):
+        """Copy data from the inputCatalog into an output catalog with
+        requested columns.
+
+        Parameters
+        ----------
+        inputCatalog: `lsst.afw.table.SourceCatalog`
+           Input catalog with data to be copied into new output catalog.
+
+        Returns
+        -------
+        outputCatalog: `lsst.afw.table.SourceCatalog`
+            Output catalog with data copied from input and new column names.
+        """
+        outputCatalog = afwTable.SourceCatalog(self.outputSchema)
+        outputCatalog.preallocate(len(inputCatalog))
+
+        for inputRecord in inputCatalog:
+            outputRecord = outputCatalog.addNew()
+            outputRecord.assign(inputRecord, self.mapper)
+
+        if outputCatalog.isContiguous():
```

Comment:
In your testing, have you found that this is needed?   

Jim says: "I also think the `isContiguous` check should be unnecessary; maybe best to turn that into an `assert` to make sure I'm not wrong about that and document-in-code that it's guaranteed."

Reviewer's personal preference is throwing exceptions instead of asserts, but coder's choice.